packages/langium/src/workspace/documents.ts

```diff
@@ -177,19 +194,61 @@ export class DefaultLangiumDocumentFactory implements LangiumDocumentFactory {
         if (uri === undefined) {
             uri = URI.parse(textDocument!.uri);
         }
-        const services = this.serviceRegistry.getServices(uri);
+
         let parseResult: ParseResult<T>;
         if (model === undefined) {
-            parseResult = services.parser.LangiumParser.parse<T>(text ?? textDocument!.getText());
+            parseResult = this.parse<T>(uri, text ?? textDocument!.getText());
         } else {
             parseResult = { value: model, parserErrors: [], lexerErrors: [] };
         }
-        return this.createLangiumDocument<T>(parseResult, uri, textDocument ?? {
+
+        return this.createLangiumDocument<T>(parseResult, uri, textDocument ?? this.createTextDocumentTemplate(uri, text!));
+    }
+
+    update<T extends AstNode = AstNode>(document: Mutable<LangiumDocument<T>>): LangiumDocument<T> {
+        const textDocument = this.textDocuments.get(document.uri.toString());
+        const text = textDocument ? textDocument.getText() : this.getContent(document.uri);
+
+        if (textDocument)
+            Object.defineProperty(
+                document, 'textDocument', {
+                    value: textDocument
+                }
+            );
+        else {
+            const documentTemplate = this.createTextDocumentTemplate(document.uri, text);
+            const textDocumentGetter = this.getTextDocumentGetter(documentTemplate);
+            Object.defineProperty(
+                document, 'textDocument', {
+                    get: textDocumentGetter
+                }
+            );
+        }
```

Comment:
This is good. Proposal: do the same in `createLangiumDocument` (i.e. define a value property if the text document is already present), then `getTextDocumentGetter` would be used only with a template.

Reply:
Actually I thought about doing that for consistency.
But in 99.9999% of the executions the 'get' variant will be used, so I skipped it.
Would you still include it, too?

Reply:
You are right, but when you get into the unlikely case it's good if the data has the same shape as always.